lib/worker.js

```diff
@@ -0,0 +1,95 @@
+'use strict'
+
+const pino = require('../pino.js')
+const { Transform, pipeline } = require('stream')
+const build = require('pino-abstract-transport')
+const { once } = require('events')
+
+// This file is not checked by the code coverage tool,
+// as it is not reliable.
+
+/* istanbul ignore file */
+
+module.exports = async function ({ destinations }) {
+  destinations = await Promise.all(destinations.map(async (t) => {
+    let stream
+    if (t.source) {
+      const toLoad = 'file://' + t.source
+      stream = await (await import(toLoad)).default(t.options)
+    } else if (t.prettyPrint) {
+      const pretty = require('pino-pretty')(t.prettyPrint)
+      stream = new Transform({
+        objectMode: true,
+        autoDestroy: true,
+        transform (chunk, enc, cb) {
+          const line = pretty(chunk.toString())
+          cb(null, line)
+        }
+      })
+
+      // TODO figure out why sync: false does not work here
+      // TODO remove the istanbul ignore, add a test using a
+      // child_process
+      const destination = pino.destination(t.destination || /* istanbul ignore next */ 1)
+      /* istanbul ignore next */
+      if (destination.fd === 1) {
+        destination.end = function () {
+          this.emit('close')
+        }
+      }
+      await once(destination, 'ready')
+      pipeline(stream, destination, () => {})
+    } else {
+      // TODO remove the transform
+      stream = new Transform({
+        objectMode: true,
+        autoDestroy: true,
+        transform (chunk, enc, cb) {
+          cb(null, chunk.toString() + '\n')
+        }
+      })
+
+      // TODO figure out why sync: false does not work here
+      const destination = pino.destination(t.destination || /* istanbul ignore next */ 1)
+      await once(destination, 'ready')
+      pipeline(stream, destination, () => {})
+    }
+    return {
+      level: t.level,
+      stream
+    }
+  }))
+  return build(process, {
+    parse: 'lines',
+    metadata: true,
+    close (err, cb) {
+      let expected = 0
+      for (const transport of destinations) {
+        expected++
+        transport.stream.on('close', closeCb)
+        transport.stream.end()
+      }
+
+      function closeCb () {
+        if (--expected === 0) {
+          cb(err)
+        }
+      }
+    }
+  })
+
+  function process (stream) {
+    const multi = pino.multistream(destinations)
+    // TODO manage backpressure
```

Comment:
can we close our this PR before this todo is done?

Reply:
This is something that we should talk about. I'm not 100% convinced we should implement this at all: a slow or stalled destination could block all log processing.

The current behavior is to have memory filled in. At some point the worker would crash, which will terminate the process. How do we want to handle this case?

Reply:
this is path way to log loss... 

can we have a fallback? log to file to clear memory?

Reply:
this is currently a problem with process transports as well though right - a solution for this would make the switch to v7 even more compelling